ui-framework-documentation.md

```diff
@@ -76,34 +89,85 @@ sub provide {
+}
+```
+The method get_partitioner_controller calls the right controller according to the product version,
+eg for Tumbleweed:
+
+
+```perl
+package Distribution::Opensuse::Tumbleweed;
+use Installation::Partitioner::LibstorageNG::v4_3::ExpertPartitionerController;
+
+sub get_expert_partitioner {
+    return Installation::Partitioner::LibstorageNG::v4_3::ExpertPartitionerController->new();
 }
 ```
+In this example, the proper controller is v4.3. The class re-uses common parts from older versions of the product.
+```
+.
+├── ExpertPartitionerPage.pm                     # Base page class
+├── FormattingOptionsPage.pm                     # Base page class 
+├── Libstorage                                      
+│   ├── ExpertPartitionerController.pm           # Base controller class
+│   ├── FormattingOptionsPage.pm                 # Inherited page class   
+├── LibstorageNG
+│   ├── ExpertPartitionerPage.pm                 # Inherited page class   
+│   ├── v3
+│   │   └── ExpertPartitionerController.pm       # Inherited controller class
+│   ├── v4
+│   │   └── ExpertPartitionerController.pm       # Inherited controller class
+│   └── v4_3
+│       ├── ExpertPartitionerController.pm       # Inherited controller class
+│       ├── ExpertPartitionerPage.pm             # Inherited page class
+│       ├── FormatMountOptionsPage.pm            # Inherited page class
+```
 
 ## Framework Layers
 
+Abstract: All direct interactions with the GUI are defined in a [page](#page), those methods are grouped in higher-level methods 
+ within a [controller](#controller) to form 
+"[business
+logic](https://en.wikipedia.org/wiki/Business_logic)" actions, and the methods from the controller are called from a
+[test module](#test-module).
+* Note: The page-object model only applies to interactions with GUI, so any interaction with the
+  command-line can take place directly in the test module and use testapi
+  directly. But the use of conditional schedule and [external test data](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/declarative-schedule-doc.md#test_data)
+  should always be considered in order to avoid "spaghetti code".
+
 ### Test Module
 
 Test Module is a layer containing test case steps that need to be
 executed on the system under test (SUT).
 
+ The variables to be used in the test are defined here before being passed to the other layers.
+In our test module, we should not care about the specifics of each SUT. If we want to format some disks,
+ We may create a method called ```format_disks```, defined in a [controller](#controller), that can be re-used 
+in all variants of (Open)Suse. It should describe a [business
+logic](https://en.wikipedia.org/wiki/Business_logic) worflow, using a verb and an object. 
+Instead of adding conditions in the module, we should use different [test data](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/declarative-schedule-doc.md#test_data) for each case,
```

Comment:
s/for each case/for each case only if the product behaves different/
(should be clear that we aim to have the same coverage regardless the product)

Reply:
I added the word "specific" to stress it, otherwise that makes the sentence very heavy and I think it's clear enough that it's only in case we would need a condition.